### PR TITLE
feat(medusa): Configurable returnable_items on order decorate totals

### PR DESCRIPTION
**Pull request by adrien2p**
**What**
Allow to request the returnable items upon decorating the order with the totals

This pr should also improve the perf around the orders as the totals is now using the new totals instead of the old one, at least for the final retrieve

Resolves CORE-1134

### CODE REVIEW

File: packages/medusa/src/api/routes/admin/orders/create-swap.ts
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:109]: It is the same approach as we have done for the total fields, but my first test was having another options in the qp which was `include_returnable_items` but since in the end it add a property on the object I wanted to avoid confusion for the users in regards to the difference between fields and the options. 

In regard to applying it on different endpoint, I did not have done it for this example but if we agree on the approach then all of that will be moved to a middleware which will assign the option on the req filterable fields

Comment (olivermrbl) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:109]: > will be moved to a middleware which will assign the option on the req filterable fields

This sounds like a solid approach to me. Would it make sense to introduce a separate query param e.g. `include` for these fields?

That way, we could have an `allowedIncludes` property similar to how we approach the relations and fields, and we don't add a property on the object.

Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:109]: When I said object I was referring to the order on which the returnable_items are attached to 😄 but yeah we can introduce an include param indeed which would work like fields or relations 👍 I just wanted to avoid confusion between fields and include* as to why for example a user has to add `total` to `fields` and something else to `include`. But at the same time, since the totals fields are disappearing from the `fields` I think it is not too much of a problem. wdyt?

Comment (olivermrbl) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:109]: Actually, after giving this some thought, I think it is reasonable to add this as an allowed field - and not use a separate property. 

How would the middleware look like for this?

Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:109]: I ll do a draft proposal asap

File: packages/medusa/src/api/middlewares/order/apply-order-includes-options.ts
Comment (adrien2p) [inline packages/medusa/src/api/middlewares/order/apply-order-includes-options.ts:14]: @olivermrbl here is the proposed middleware

Comment (olivermrbl) [reply packages/medusa/src/api/middlewares/order/apply-order-includes-options.ts:14]: comment: I like the idea. As you've alright highlighted, this is unnecessarily coupled with Orders. The `includes` fields can be defined in the `index.ts` of domains and passed to the middleware when registering it. Similar to our transformers.

File: packages/medusa/src/api/routes/admin/orders/create-swap.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/orders/create-swap.ts:104]: thought(non-blocking): Instead of adding clutter to our handlers, could we pass `...req.includes` to the retrieve options?

File: packages/medusa/src/api/middlewares/order/apply-order-includes-options.ts
Comment (adrien2p) [reply packages/medusa/src/api/middlewares/order/apply-order-includes-options.ts:14]: I will proceed with the changes then 👍

File: packages/medusa/src/api/routes/admin/orders/create-swap.ts
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:104]: I think so, but we can imagine that includes is not necessarily link to the totalFieldsConfig and can contains something else. So I would go for a more verbose but more clear (and people knows what to expect) instead. wdyt?

Comment (olivermrbl) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:104]: One could argue that having a dedicated config for totals is not ideal. This could be a `context` instead.

That said, it's a tradeoff between verbosity and flexibility, and I am not sure which one is more appropriate. What's your gut-feeling?

Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/create-swap.ts:104]: I agree, I will just change the name of the parameter to context and add an includes property which would take the includes object. 👍

**Comment by olivermrbl** [comment]
@kasperkristensen Do you have any feedback on this approach?

File: packages/medusa/src/api/routes/admin/orders/create-swap.ts
Comment (patrick-medusajs) [inline packages/medusa/src/api/routes/admin/orders/create-swap.ts:34]: todo: update OAS `x-codegen` with:
```
x-codegen:
  method: createSwap
  queryParams: AdminPostOrdersOrderSwapsParams
```

File: packages/medusa/src/models/order.ts
Comment (patrick-medusajs) [inline packages/medusa/src/models/order.ts:553]: todo?: I believe we are missing:
```
Available if the relation `returnable_items ` is expanded.
```

**Review by patrick-medusajs — COMMENTED** [review]
Small todos regarding the OAS.

File: packages/medusa/src/models/order.ts
Comment (adrien2p) [reply packages/medusa/src/models/order.ts:553]: There is no relation for that, it is a computed fields like the other totals 👍

**Review by patrick-medusajs — APPROVED** [review]
LGTM
Praise for converting the snapshot test along the way.

**Review by olivermrbl — CHANGES_REQUESTED** [review]
This is great @adrien2p!

Please add the middleware to all `/admin/orders` endpoints :)

**Comment by adrien2p** [comment]
> This is great @adrien2p!
> 
> Please add the middleware to all `/admin/orders` endpoints :)

I will that now so that we can then merge 👍

**Review by kasperkristensen — APPROVED** [review]
LGTM, really great addition that will make RMA much easier to handle moving forward 🚀

File: packages/medusa/src/api/routes/admin/orders/index.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/orders/index.ts:300]: todo: I wouldn't make this a default field. The returnable items are only relevant in RMA flows right now, so no need to do add more load than needed on all other calls.

File: packages/medusa/src/services/order.ts
Comment (olivermrbl) [inline packages/medusa/src/services/order.ts:1791]: todo: I believe we should make `item.shipped_quantity` instead, as you can cancel a fulfillment from admin instead of returning it.

**Review by olivermrbl — APPROVED** [review]
LGTM! Just a couple of todos before we merge :)

**Comment by adrien2p** [comment]
> This is great @adrien2p!
> 
> Please add the middleware to all `/admin/orders` endpoints :)

@olivermrbl This is done and I did some scout as well

File: packages/medusa/src/api/middlewares/transform-includes-options.ts
Comment (olivermrbl) [inline packages/medusa/src/api/middlewares/transform-includes-options.ts:10]: thought: Wouldn't it be better to default to an empty array here instead of having to remember to do: `transformIncludesOptions([])`?

File: packages/medusa/src/api/routes/admin/orders/process-swap-payment.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/orders/process-swap-payment.ts:73]: question: No totals?

**Comment by olivermrbl** [comment]
Impeccable boyscouting here! Just need tests to pass :)

File: packages/medusa/src/api/routes/admin/orders/process-swap-payment.ts
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/process-swap-payment.ts:73]: Yeah apparently we were not returning the total here but I can add it :)

File: packages/medusa/src/models/order.ts
Comment (adrien2p) [reply packages/medusa/src/models/order.ts:553]: already done locally :D thanks man

**Comment by patrick-medusajs** [comment]
LGTM
I re-reviewed the OAS with the latest changed. Thanks 🙌

File: packages/medusa/src/api/routes/admin/orders/index.ts
Comment (olivermrbl) [reply packages/medusa/src/api/routes/admin/orders/index.ts:509]: Looks like this was missed in the rebase. Should not be a default field 👍

File: packages/medusa/src/api/routes/admin/orders/index.ts
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/index.ts:6]: 😄 yes indeed, finally

**Review by olivermrbl — APPROVED** [review]
Great job! Let's await @kasperkristensen's approval to see if something comes up. He is currently incorporating it in the RMA flows 👍

**Comment by kasperkristensen** [comment]
@adrien2p Not sure if it is related to me having the usual issues with `medusa-dev` being a bit finicky, but I can't get this to work in our admin. 

When passing `returnable_items` to `fields` in `useAdminOrder` the backend throws the following error every time:

```
error:   returnable_items column was not found in the Order entity.
TypeORMError: returnable_items column was not found in the Order entity.
```

**Comment by adrien2p** [comment]
> @adrien2p Not sure if it is related to me having the usual issues with `medusa-dev` being a bit finicky, but I can't get this to work in our admin.
> 
> When passing `returnable_items` to `fields` in `useAdminOrder` the backend throws the following error every time:
> 
> ```
> error:   returnable_items column was not found in the Order entity.
> TypeORMError: returnable_items column was not found in the Order entity.
> ```

@kasperkristensen, I ve added it to the missing route [here](https://github.com/medusajs/medusa/pull/3276/commits/ffcabb933be2754cecec0f1ad9d1cd1651de5a3f) The next snapshot should include it 🚀

**Review by kasperkristensen — APPROVED** [review]
LGTM! I just wrapped up using it in the admin and it makes it so simple to display RMA flows correctly with minimal calculations on the frontend. 

One very minor nit (non-blocking): Using the `fields` parameter requires the consumer to also pass all the default fields, as simply passing `fields: returnable_items` results in a very bare-bones order being returned. This isn't the greatest DX, as you will need to dig into our code base to figure out which fields to re-add in order to get the order object you previously received + returnable items.

**Comment by kasperkristensen** [comment]
It works @adrien2p 💪

**Comment by adrien2p** [comment]
@olivermrbl you should be able to re test to get an order with `fields=id&expand` and the result should be the expected one

**Comment by olivermrbl** [comment]
@adrien2p Did you test this? I am now facing the same issue as reported earlier:
```
error:   returnable_items column was not found in the Order entity.
TypeORMError: returnable_items column was not found in the Order entity.
```

**Comment by olivermrbl** [comment]
Also, I have differentiating behaviour on listing orders vs get order:
```
GET /admin/orders/:id?fields=id
{ order: { id: "some" } }
```

```
GET /admin/orders?fields=id
{ order: { id: "some", ...[all relations] } }
```

**Comment by adrien2p** [comment]
> @adrien2p Did you test this? I am now facing the same issue as reported earlier:
> 
> ```
> error:   returnable_items column was not found in the Order entity.
> TypeORMError: returnable_items column was not found in the Order entity.
> ```

@olivermrbl of course I tested it 😂 
Your error seams weird since the only change is related to the response and nothing else. Are you sure you have the latest version. Have you pulled the changes or are you a snapshot?

**Comment by olivermrbl** [comment]
> Have you pulled the changes or are you using a snapshot

I pulled the branch. Checked the `dist` and they contain your latest changes. Let's try with a snapshot, but I am dreading it wont make a difference 😊

**Comment by adrien2p** [comment]
Screenshot 2023-02-21 at 10 46 53
Screenshot 2023-02-21 at 10 47 12

**Pull request merged**
